Index: pyproject.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>[build-system]\nrequires = [\"hatchling\"]\nbuild-backend = \"hatchling.build\"\n\n[project]\nname = \"PyOghma\"\nversion = \"0.0.12\"\nauthors = [\n  { name=\"Cai Williams\", email=\"cai.williams@physik.tu-chemnitz.de\" },\n]\ndescription = \"A Python API for OghmaNano\"\nreadme = \"README.md\"\ndependencies = [\"mgzip>=0.2.1\", \"numpy>=1.26.4\", \"pandas>=2.2.2\", \"tqdm>=4.66.2\", \"ujson>=5.9.0\", ]\n\n[project.urls]\nHomepage = \"https://github.com/CaiWilliams/pyOghma\"\nIssues = \"https://github.com/CaiWilliams/pyOghma/issues\"
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pyproject.toml b/pyproject.toml
--- a/pyproject.toml	(revision 5542ccf1cb68122a94acf21c65a1f3b614fbbe49)
+++ b/pyproject.toml	(date 1721290911817)
@@ -4,7 +4,7 @@
 
 [project]
 name = "PyOghma"
-version = "0.0.12"
+version = "0.0.16"
 authors = [
   { name="Cai Williams", email="cai.williams@physik.tu-chemnitz.de" },
 ]
Index: .idea/material_theme_project_new.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+><?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<project version=\"4\">\n  <component name=\"MaterialThemeProjectNewConfig\">\n    <option name=\"metadata\">\n      <MTProjectMetadataState>\n        <option name=\"userId\" value=\"-45b10606:190ba95cb1f:-7fcb\" />\n      </MTProjectMetadataState>\n    </option>\n  </component>\n</project>
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/material_theme_project_new.xml b/.idea/material_theme_project_new.xml
--- a/.idea/material_theme_project_new.xml	(revision 5542ccf1cb68122a94acf21c65a1f3b614fbbe49)
+++ b/.idea/material_theme_project_new.xml	(date 1721292744847)
@@ -3,7 +3,9 @@
   <component name="MaterialThemeProjectNewConfig">
     <option name="metadata">
       <MTProjectMetadataState>
-        <option name="userId" value="-45b10606:190ba95cb1f:-7fcb" />
+        <option name="migrated" value="true" />
+        <option name="pristineConfig" value="false" />
+        <option name="userId" value="6153fc75:19039d087a1:-7ff9" />
       </MTProjectMetadataState>
     </option>
   </component>
